firedrake/interpolation.py

```diff
@@ -227,17 +242,55 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
         raise RuntimeError('Shape mismatch: Expression shape %r, FunctionSpace shape %r'
                            % (expr.ufl_shape, V.ufl_element().value_shape()))
 
-    mesh = V.ufl_domain()
-    coords = mesh.coordinates
+    target_mesh = V.ufl_domain()
+    try:
+        trans_mesh = (
+            expr.ufl_domain() is not None  # Constant expr
+            and expr.ufl_domain() != V.mesh()  # Coming from a different domain
+        )
+    except AttributeError:
+        # Have python Expression
```

Comment:
It looks this is new in this PR?

Reply:
It's not - python expressions are allowed (see for example this test from `tests/regression/test_python_parloop.py` 
```python
def test_python_parloop():

    m = UnitSquareMesh(4, 4)
    fs = FunctionSpace(m, "CG", 2)
    f = Function(fs)

    class MyExpression(Expression):
        def eval(self, value, X):

            value[:] = np.dot(X, X)

    f.interpolate(MyExpression())
    X = m.coordinates
    assert assemble((f-dot(X, X))**2*dx)**.5 < 1.e-15
```
but these expressions throw an `AttributeError` if you try to access their `ufl_domain` property. This allows for that to happen.

Reply:
I see. Then we can probably override `ufl_domain` method of `firedrake.Expression` so that it would always return `None`?

Reply:
Yes but I'm not going to do that in this PR